lib/matplotlib/tight_layout.py

```diff
@@ -81,23 +60,19 @@ def auto_adjust_subplotpars(
         margin_left = margin_bottom = margin_right = margin_top = None
     else:
         margin_left, margin_bottom, _right, _top = rect
-        if _right:
-            margin_right = 1 - _right
-        else:
-            margin_right = None
-        if _top:
-            margin_top = 1 - _top
-        else:
-            margin_top = None
+        margin_right = 1 - _right if _right else None
+        margin_top = 1 - _top if _top else None
 
-    vspaces = [[] for _ in range((rows + 1) * cols)]
-    hspaces = [[] for _ in range(rows * (cols + 1))]
-
-    union = Bbox.union
+    vspaces = np.full((rows + 1, cols), None)
```

Comment:
The third dimension is `len(zip(subplot_list, ax_bbox_list, num1num2_list))` minus the `subplots` with completely invisible Axes. But since we take the sum of the lists in `vspaces`/`hspaces`, then we can just fill the empty locations with 0, and make these truly 3D?

Reply:
Actually we can just directly accumulate the values (it's not clear to me why you'd want to *add* paddings when multiple axes share the same subplotspec (which the comment says is possible in axes_grid?? well let's not get into that for now), but let's not change that here).
The last dim is not actyally len(zip(...)), but the max number of axes that share the same subplotspec.